src/rtc.rs

```diff
@@ -0,0 +1,448 @@
+        let mut result = Self {
+            regs,
+            prediv_s,
+            prediv_a,
+        };
+
+        enable_lse(bdcr, false);
+        unlock(apb1);
+        enable(bdcr);
+        result.set_24h_fmt();
+
+        result.regs.prer.modify(|_, w| unsafe {
+            w.prediv_s().bits(prediv_s);
+            w.prediv_a().bits(prediv_a)
+        });
+
+        result
+    }
+
+    /// Sets calendar clock to 24 hr format
+    pub fn set_24h_fmt(&mut self) {
+        self.regs.cr.modify(|_, w| w.fmt().set_bit());
+    }
+    /// Sets calendar clock to 12 hr format
+    pub fn set_12h_fmt(&mut self) {
+        self.regs.cr.modify(|_, w| w.fmt().clear_bit());
+    }
+
+    /// Reads current hour format selection
+    pub fn is_24h_fmt(&self) -> bool {
+        self.regs.cr.read().fmt().bit()
+    }
+
+    /// As described in Section 27.3.7 in RM0316,
+    /// this function is used to disable write protection
+    /// when modifying an RTC register
+    fn modify<F>(&mut self, mut closure: F)
+    where
+        F: FnMut(&mut RTC),
+    {
+        // Disable write protection
+        self.regs.wpr.write(|w| unsafe { w.bits(0xCA) });
+        self.regs.wpr.write(|w| unsafe { w.bits(0x53) });
+        // Enter init mode
+        let isr = self.regs.isr.read();
+        if isr.initf().bit_is_clear() {
+            self.regs.isr.modify(|_, w| w.init().set_bit());
+            while self.regs.isr.read().initf().bit_is_clear() {}
+        }
+        // Invoke closure
+        closure(&mut self.regs);
+        // Exit init mode
+        self.regs.isr.modify(|_, w| w.init().clear_bit());
+        // wait for last write to be done
+        while !self.regs.isr.read().initf().bit_is_clear() {}
+    }
+}
+
+impl Rtcc for Rtc {
+    type Error = Error;
+
+    /// set time using NaiveTime (ISO 8601 time without timezone)
+    /// Hour format is 24h
+    fn set_time(&mut self, time: &NaiveTime) -> Result<(), Self::Error> {
+        self.set_24h_fmt();
+        let (ht, hu) = bcd2_encode(time.hour());
+        let (mnt, mnu) = bcd2_encode(time.minute());
+        let (st, su) = bcd2_encode(time.second());
+        self.modify(|regs| {
+            regs.tr.write(|w| unsafe {
+                w.ht()
+                    .bits(ht)
+                    .hu()
+                    .bits(hu)
+                    .mnt()
+                    .bits(mnt)
+                    .mnu()
+                    .bits(mnu)
+                    .st()
+                    .bits(st)
+                    .su()
+                    .bits(su)
+                    .pm()
+                    .clear_bit()
+            });
+        });
+
+        Ok(())
+    }
+
+    fn set_seconds(&mut self, seconds: u8) -> Result<(), Self::Error> {
+        if seconds > 59 {
+            return Err(Error::InvalidInputData);
+        }
+        let (st, su) = bcd2_encode(seconds as u32);
+        self.modify(|regs| regs.tr.write(|w| unsafe { w.st().bits(st).su().bits(su) }));
```

Comment:
Using a `write` will reset all the fields not written. I believe in this case this means all other time fields (minutes, hours, ...) will be set to 0. I guess that is not the intended behavior?

Same for some of the other `set_*` methods below. I suggest you check all register `write`s again and make sure they shouldn't actually be `modify`s.